rest/route/distro.go

```diff
@@ -2,14 +2,114 @@ package route
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
 	"net/http"
+	"strings"
 
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/distro"
 	"github.com/evergreen-ci/evergreen/rest/data"
 	"github.com/evergreen-ci/evergreen/rest/model"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/evergreen-ci/evergreen/validator"
 	"github.com/evergreen-ci/gimlet"
 	"github.com/pkg/errors"
 )
 
+////////////////////////////////////////////////////////////////////////
+//
+// PATCH /rest/v2/distros/{distro_id}
+
+type distroIDPatchHandler struct {
+	distroId string
+	body     []byte
+	sc       data.Connector
+}
+
+func makePatchDistroByID(sc data.Connector) gimlet.RouteHandler {
+	return &distroIDPatchHandler{
+		sc: sc,
+	}
+}
+
+func (h *distroIDPatchHandler) Factory() gimlet.RouteHandler {
+	return &distroIDPatchHandler{
+		sc: h.sc,
+	}
+}
+
+// Parse() fetches the distroId and json payload from the http request.
+func (h *distroIDPatchHandler) Parse(ctx context.Context, r *http.Request) error {
+	h.distroId = gimlet.GetVars(r)["distro_id"]
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	b, err := ioutil.ReadAll(body)
+	if err != nil {
+		return errors.Wrap(err, "Argument read error")
+	}
+	h.body = b
+
+	return nil
+}
+
+// Run() finds a distro by id; validates its patched state, which is updated
+// and returned if valid
+func (h *distroIDPatchHandler) Run(ctx context.Context) gimlet.Responder {
+	d, err := h.sc.FindDistroById(h.distroId)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "Database error"))
+	}
+
+	apiDistro := &model.APIDistro{}
+	if err = apiDistro.BuildFromService(*d); err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "API model error"))
+	}
+
+	if err = json.Unmarshal(h.body, apiDistro); err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "API model error"))
+	}
+
+	i, err := apiDistro.ToService()
+	if err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "API model error"))
+	}
+	d = i.(*distro.Distro)
+
+	if d.Id != h.distroId {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusBadRequest,
+			Message:    fmt.Sprintf("Distro name is immutable; cannot rename distro '%s'", h.distroId),
+		})
+	}
+
+	vErrors, err := validator.CheckDistro(ctx, d, &evergreen.Settings{}, false)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusBadRequest,
+			Message:    "Database error",
+		})
+	}
+	if len(vErrors) != 0 {
+		var m strings.Builder
```

Comment:
I've never used a strings.Builder! This looks cool (I think it's possible that this is too new for the version of go that we run tests on, but it's fine to just build a list and use strings.Join()